Sources/Bodega/StorageEngine.swift

```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+/// A `StorageEngine` represents an underlying data store mechanism for saving and persisting data.
+///
+/// A ``StorageEngine`` is a construct you can build that plugs into ``ObjectStorage``
+/// to use for persisting data.
+///
+/// This library has two implementations of `StorageEngine`, ``DiskStorageEngine`` and ``SQLiteStorageEngine``.
+/// Both of these can serve as inspiration if you have your own persistence mechanism (such as Realm, CoreData, etc).
+///
+/// `DiskStorageEngine` takes `Data` and saves it to disk using file system operations.
+/// `SQLiteStorageEngine` takes `Data` and saves it to an SQLite database under the hood.
+///
+/// If you have your own way of storing data then you can refer to `DiskStorageEngine` and `SQLiteStorageEngine`
+/// for inspiration, but all you need to do is conform to the `StorageEngine` protocol
+/// and initialize `ObjectStorage` with that storage.
+public protocol StorageEngine: Actor {
+    var directory: FileManager.Directory { get }
+
+    func write(_ data: Data, key: CacheKey) throws
+    func write(_ dataAndKeys: [(key: CacheKey, data: Data)]) throws
+
+    func read(key: CacheKey) -> Data?
```

Comment:
You can relieve some burden from conformers by providing default implementations when you have `foo(_ value: Foo)` and `foo(_ value: [Foo])` (i.e. singular/plural instances)

i.e. 
```swift
extension StorageEngine {
    func read(key: CacheKey) -> Data? {
        return read(keys: [key])?.first
    }
}
```

You can do this for `read`, `remove` etc..

Keeping them in the protocol however will leave it open for customization in the event performance optimizations are available (depending on the underlying implementation)